#26121 - special price & tier price are coming in base currency

pmarjan opened the pull request:
Collect store specific value for store specific currecny when querying products special_price and price_tiers

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
Solves for https://github.com/magento/magento2/issues/26121

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Collect store specific value for store specific currecny when querying products special_price and price_tiers

pmarjan commented:
@magento run all tests

danielrenaud commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,42 @@
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+use Magento\Catalog\Pricing\Price\SpecialPrice as PricingSpecialPrice;
+
+class SpecialPrice implements ResolverInterface
```
Class requires a short description

danielrenaud commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,42 @@
+     * @param array|null $value
+     * @param array|null $args
+     * @return mixed|Value
+     * @throws \Exception
```
This should not throw \Exception. Resolvers should only throw graphql specific exceptions

danielrenaud commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,42 @@
+     * @param array|null $args
+     * @return mixed|Value
+     * @throws \Exception
+     */public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
```
look like the comment closing tag is on same line as function

cpartica replied:
static tests fail, some params are not used

danielrenaud reviewed (changes requested):
Can you please update or add test automated tests to cover these changes.

cpartica commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,42 @@
+class SpecialPrice implements ResolverInterface
+{
+    /**
+     * Fetches the data from persistence models and format it according to the GraphQL schema.
```
Fetch the data (Imperative, Fetches would be descriptive)

cpartica commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Fetches the data from persistence models and format it according to the GraphQL schema.
+     *
+     * @param \Magento\Framework\GraphQl\Config\Element\Field $field
```
move FQDN to use statement

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Adjust to take customer group (via setting customerGroupId) into consideration.

pmarjan commented:
@magento run all tests

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Edits, Support by api-functional tests, Address static tests

pmarjan commented:
@magento run all tests

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Address static tests - session 2

pmarjan commented:
@magento run all tests

pmarjan commented:
@cpartica all points from above should be addressed now. Also, the resolvers for special price and tier prices are supported now by api-functional tests that specifically query for a second store that has a display currency that comes with some rate compared to the base currency.

danielrenaud commented on app/code/Magento/Catalog/Model/Product/Type/Price.php:
```diff
@@ -368,10 +368,33 @@ protected function getAllCustomerGroupsId()
+     * @param array $tierPricesRaw
+     * @return \Magento\Catalog\Api\Data\ProductTierPriceInterface[]
+     */
+    public function transformTierPrices($tierPricesRaw)
```
Adding a public method to an "@api" class is a backwards incompatible change, that is best avoided.

cpartica replied:
I tink the alternative is to extract \Magento\Catalog\Model\Product\Type\Price::buildProductTierPriceInterfaceObjects and \Magento\Catalog\Model\Product\Type\Price::getWebsiteForPriceScope into another class..
we can deprecate getWebsiteForPriceScope and buildProductTierPriceInterfaceObjects

Later in 2.5.0 we can delete them.
in 2.4.1 in theory we can't even add this new class to the constructor with "= null" Backward compatible way unless we have an approval. we can try.
Anyway either that or code duplication.

pmarjan replied:
For this purpose, created now is Magento\Catalog\Model\Product\Price\TierPriceBuilder.

danielrenaud commented on app/code/Magento/CatalogCustomerGraphQl/Pricing/Price/TierPrice.php:
```diff
@@ -0,0 +1,47 @@
+use Magento\Framework\Pricing\PriceCurrencyInterface;
+use Magento\Framework\App\ObjectManager;
+
+class TierPrice extends \Magento\Catalog\Pricing\Price\TierPrice
```
We should avoid doing extension in this way.
I necessary it may be better to add customerGroupId as an optional param for methods, and falling back to the old mechanism when not provided

pmarjan replied:
I remember, we cannot use \Magento\Catalog\Pricing\Price\TierPrice directly (even if we would apply proposed chances that would properly set the customer group).

Reason is that the mess detector test will fail with
`The class Magento\Catalog\Pricing\Price\TierPrice uses sessions or cookies while not being a part of HTML Presentation layer`
Unfortunately, this class has the session obsolete and not used at all, but still, cannot be directly called in graphql area.

Anyway, my latest commit bypasses this idea completely, and this should not be an issue anymore.

danielrenaud commented on app/code/Magento/CatalogCustomerGraphQl/Model/Resolver/Product/Price/Tiers.php:
```diff
@@ -97,7 +115,23 @@ public function getProductTierPrices($productId): ?array
             return null;
         }
-        return $this->products[$productId]->getTierPrices();
+
```
Is there a reason that the original Price::getTierPrices method does not for us here?

pmarjan replied:
this original Price::getTierPrices actually loads always in base currency not minding the current store selected currency.
Now, trying to see what can be done about it - maybe this method can be improved.

danielrenaud reviewed (changes requested):
I left some comments that should be addressed, but we can discuss if anything is unclear

cpartica commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,36 @@
+        /** @var \Magento\Catalog\Model\Product $product */
+        $product = $value['model'];
+        /** @var PricingSpecialPrice $specialPrice */
+        $specialPrice = $product->getPriceInfo()->getPrice(PricingSpecialPrice::PRICE_CODE);
```
you already know how to get here customerGroup, so use it here to get the price you want, pass it by param

cpartica commented on app/code/Magento/Catalog/Model/Product/Type/Price.php:
```diff
@@ -368,10 +368,33 @@ protected function getAllCustomerGroupsId()
+     * Return ProductTierPriceInterface[] given raw tier prices array
+     *
+     * @param array $tierPricesRaw
+     * @return \Magento\Catalog\Api\Data\ProductTierPriceInterface[]
```
can you move this to use?

cpartica commented on app/code/Magento/CatalogGraphQl/Model/Resolver/Product/SpecialPrice.php:
```diff
@@ -0,0 +1,36 @@
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
+    {
+        /** @var \Magento\Catalog\Model\Product $product */
```
use product interface

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Refactor to pass new optional parameter to @api class constructor

pmarjan commented:
@magento run all tests

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Refactor TierPriceBuilder

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Revert Pricing Price TierPrice

pmarjan commented:
@magento run all tests

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Adjust unit and integration tests

pmarjan commented:
@magento run all tests

pmarjan commented:
@cpartica @danielrenaud 
all remarks are addressed, PR is ready for new re-review

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Move price transform to PriceTiers

pmarjan commented:
@magento run all tests

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+     */
+    public function testAllGroups()
+    {
+        /** @var string $productSku */
```
```suggestion
```
Remove all unnecessary @var annotations

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+        $response = $this->graphQlQuery($query);
+
+        $itemTiers = $response['products']['items'][0]['price_tiers'];
+        $this->assertEquals(5, sizeof($itemTiers));
```
You can use assertCount()

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+     * @param string $password
+     * @return array
+     */
+    private function getHeaderAuthorization(string $username, string $password): array
```
Please use \Magento\GraphQl\GetCustomerAuthenticationHeader instead

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+        $productRepository = $this->objectManager->get(ProductRepositoryInterface::class);
+        /** @var Product $product */
+        $product = $productRepository->get($productSku, false, null, true);
+        $tierPriceData =[
```
data changes should be done in fixtures

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+        );
+
+        $itemTiers = $response['products']['items'][0]['price_tiers'];
+        $this->assertEquals(3, sizeof($itemTiers));
```
use assertCount()

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+     */
+    public function testSecondStoreViewWithCurrencyRate()
+    {
+        /** @var string $storeViewCode */
```
```suggestion
```
Remove all unnecessary @var annotations

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+        ];
+
+        $this->saveTierPrices($product, $tierPriceData);
+        /** @var string $query */
```
```suggestion
```

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+        $productSku = 'simple';
+        /** @var string $query */
+        $query = $this->getProductSearchQuery($productSku);
+        /** @var array $headers */
```
```suggestion
```

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+     * @param array $tiers
+     * @return float
+     */
+    private function getValueForQuantity(float $quantity, array $tiers)
```
Please add short description for utility methods

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/PriceTiersTest.php:
```diff
@@ -0,0 +1,230 @@
+    {
+        return <<<QUERY
+{
+  products(search: "{$productSku}") {
```
It might be safer to use "filter sku equals" to make sure you get an exact match

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/SpecialPriceTest.php:
```diff
@@ -0,0 +1,99 @@
+     */
+    public function testSpecialPrice()
+    {
+        /** @var string $productSku */
```
```suggestion
```

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/SpecialPriceTest.php:
```diff
@@ -0,0 +1,99 @@
+
+        $response = $this->graphQlQuery($query);
+
+        /** @var float $specialPrice */
```
```suggestion
```

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/CatalogCustomer/SpecialPriceTest.php:
```diff
@@ -0,0 +1,99 @@
+    {
+        /** @var string $productSku */
+        $productSku = 'simple';
+        /** @var string $query */
```
```suggestion
```

danielrenaud reviewed (changes requested):
Functionality looks good, just some cleanup comments for the tests

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Adjustments to api-functional test

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Adjustments to api-functional test

pmarjan committed:
magento/magento2#26121: special price & tier price are coming in base currency

Adjustments to api-functional test

pmarjan commented:
@magento run all tests

danielrenaud reviewed (approved):
Nice work!
